### PR TITLE
feat(core/dfn): add data-export by default

### LINKED ISSUE #685: Implement bikeshed's dfn contract
The contract is as follows: 

- [X] The defining element MUST be a dfn or h2...h6. No other element is recognized as defining a term.
- [X] The element MUST have an id attribute.
- [x] The linking text defaults to the text content of the dfn/heading. If the desired text content isn’t suitable for linking text, or you wish to provide multiple linking texts, a `lt` attribute` containing one or more pipe-separated linking texts will override the text content. See also §6.3 Defining Extra-Short "Local" Linking Texts

Note: we don't support `local-lt`. We also "dirty" the `data-lt` with various things, like plurals... we might need to deal with that.  

- [x] `data-dfn-type` MUST be provided, and set one of the [accepted values](https://tabatkins.github.io/bikeshed/#dfn-types). (The dfn element, specifically, actually allows this to be omitted, and defaults to the "dfn" type. But headings require a type, and dfns are clearer with it specified.)

Note: it's still unclear what the heading types should be. We need to. look that up. 

- [x] Either `data-export` or `data-noexport` MAY be provided (both boolean attributes). If neither is provided, "dfn" type definitions default to noexport, while all others default to export. Unexported definitions aren’t linkable by default.

- [x] Several types of definitions are namespaced to another construct; for example, attribute names are namespaced to an interface. These definitions MUST contain a data-dfn-for attribute, containing a comma-separated list of one or more definitions they’re namespaced to.

Note:  we. can support for on IDL definitions easily enough. We need to support them for internal slots too, and possibly other other things. However, we generally don't want to export internal slots, because they are private.

### PR DESCRIPTION
**Pull request by saschanaz**
Closes #685

### CODE REVIEW

File: src/core/dfn.js [hunk 1]
```diff
@@ -17,6 +20,10 @@ export function run() {
     if (dfn.dataset.dfnFor) {
       dfn.dataset.dfnFor = dfn.dataset.dfnFor.toLowerCase();
     }
+    if (!("noexport" in dfn.dataset)) {
+      dfn.dataset.export = "";
```
Comment (sidvishnoi) [inline src/core/dfn.js]: Should we do `dfn.removeAttribute("data-export")` instead?

Comment (sidvishnoi) [reply src/core/dfn.js]: Oh. I misunderstood.

Comment (saschanaz) [reply src/core/dfn.js]: This is not removing an attribute, this is adding one.

**Comment by sidvishnoi** [comment]
I'm not clear why this is needed. Can you help me understand @saschanaz ?

**Comment by saschanaz** [comment]
@sidvishnoi To mark L2127 with `1`: https://github.com/w3c/respec/issues/2377#issuecomment-497917390

This follows export-by-default behavior, because otherwise we'll have to modify all spec documents to export something.

**Comment by marcoscaceres** [comment]
The BikeShed/Shepherd contract requires no export by default. The reason is that Editors need to think really carefully about what definitions they export and why. The only exception to this is IDL definitions, which have both a “type” and “for”.

**Comment by marcoscaceres** [comment]
I think we should do the contract incrementally: let’s start by exporting the IDL things. That would be a massive win already.

**Comment by saschanaz** [comment]
>The BikeShed/Shepherd contract requires no export by default

[You liar](https://github.com/w3c/respec/issues/685#issuecomment-487366690) 😭🤣

> let’s start by exporting the IDL things

I'll do that happily.

**Comment by marcoscaceres** [comment]
> You liar 😭🤣

heh, sorry about that... "trust, but verify".

**Comment by saschanaz** [comment]
@marcoscaceres Safe to be closed, right?

**Comment by marcoscaceres** [comment]
yes, I think so.

**Pull request closed without merging**